### PR TITLE
Add table that summarizes all functions in documentation API

### PR DESCRIPTION
**Pull request by gautierronan**
(with clickable links that bring to the appropriate function definition)

Screenshot 2023-06-23 at 13 51 37

### CODE REVIEW

File: docs/templates/python/material/module.html
Comment (pierreguilmin) [inline docs/templates/python/material/module.html:1]: When copy-pasting templates that you then modify, it'd be nice to have a commit where you just copy-paste the template, and then a subsequent commit where we see what you modify. It's a bit hard to track what is changed here.

File: dynamiqs/utils/_wigner.py
Comment (pierreguilmin) [inline dynamiqs/utils/_wigner.py:1]: I don't think we should perform this move, it's better structured like it was done, and then we can set the API the way we prefer.

**Review by pierreguilmin — CHANGES_REQUESTED** [review]
IMO this is not how we should document the API. When documenting an API the table of content should follow closely the package structure, as was done before. It's a part of the documentation that is meant for developers, not users. I'm guessing you're copying what's done in PyTorch doc, but it's a very different package with tons of functions (thus they have to structure the API somehow otherwise it's unreadable), but it's not common to do it like this, especially for small packages.

I like the table however, I think it would be better placed in a getting started page "Utils function" which gathers all utils functions we have for a quick lookup.

File: docs/templates/python/material/module.html [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+      {% endwith %}
+
+      {% with obj = module %}
+        {% set root = False %}
+        {% set heading_level = heading_level + 1 %}
+        {% if config.table %}
+          {% include "table.html" with context %}
+        {% else %}
+          {% include "children.html" with context %}
+        {% endif %}
```
Comment (gautierronan) [inline docs/templates/python/material/module.html:62]: NB: Here is the modification compared to the copy-pasted file.

File: docs/templates/python/material/module.html
Comment (gautierronan) [reply docs/templates/python/material/module.html:1]: Sorry ! I knew you'd say that but I didn't know how to backtrack through my commits ^^. I essentially just modified lines 58 > 62.

File: dynamiqs/utils/_wigner.py
Comment (gautierronan) [reply dynamiqs/utils/_wigner.py:1]: Yes I completely agree. But how can we make the functions appear on the same page with mkdocs? It seems kinda silly to have a separate page just for a single function `dq.wigner` no?

**Comment by gautierronan** [comment]
> IMO this is not how we should document the API. When documenting an API the table of content should follow closely the package structure, as was done before. It's a part of the documentation that is meant for developers, not users. I'm guessing you're copying what's done in PyTorch doc, but it's a very different package with tons of functions (thus they have to structure the API somehow otherwise it's unreadable), but it's not common to do it like this, especially for small packages.
I like the table however, I think it would be better placed in a getting started page "Utils function" which gathers all utils functions we have for a quick lookup.

Yes I really like having a table that summarizes all available functions (really missing in the qutip doc IMO). I don't really care where it is, but it should be somewhere. Let's just keep the table implementation in this PR, and think later about how we structure the doc.

File: dynamiqs/utils/_wigner.py
Comment (pierreguilmin) [reply dynamiqs/utils/_wigner.py:1]: No, the API is not a documentation for the user, it's a documentation for the developers, so it should mimic the package structure.

File: mkdocs.yml [hunk 2]
```diff
@@ -80,7 +81,11 @@ nav:
...
+  - Python API:
       - reference/index.md
-      - dynamiqs.utils:
-          - utils: reference/utils/utils.md
+      - Utility functions:
+          - Initializing states: reference/utils/states.md
+          - Initializing operators: reference/utils/operators.md
+          - Operating on states: reference/utils/utils.md
+          - Converting tensors: reference/utils/tensor_types.md
+          - Wigner distributions: reference/utils/wigner.md
```
[the remark this answers is not in the document]
Comment (pierreguilmin) [reply mkdocs.yml:91]: To be clear, we will make a user-facing documentation with all the titles you propose, but it's not what's typically done for the API documentation.

**Review by pierreguilmin — APPROVED** [review]

**Pull request merged**